### PR TITLE
rcal-687 Initial code for the high level pipeline

### PR DESCRIPTION
**Pull request by ddavis-stsci**
Resolves [RCAL-687](https://jira.stsci.edu/browse/rcal-687)

Closes https://github.com/spacetelescope/romancal/issues/924

This PR adds the level 3 pipeline code for high level processing. Testing the steps will need new data from INS.
Once we have the data the documentation and examples will be added to rtd.

The command to run the pipeline
strun --disable-crds-steppar roman_hlp r00005-a3001_image_001_short_asn.json
2023-11-14 14:34:26,342 - stpipe - WARNING - **WARNING**: LOCAL JWST PRD VERSION PRDOPSSOC-062 CANNOT BE CHECKED AGAINST ONLINE VERSION
2023-11-14 14:34:26,386 - stpipe.HighLevelPipeline - INFO - HighLevelPipeline instance created.
2023-11-14 14:34:26,387 - stpipe.HighLevelPipeline.skymatch - INFO - SkyMatchStep instance created.
2023-11-14 14:34:26,388 - stpipe.HighLevelPipeline.outlier_detection - INFO - OutlierDetectionStep instance created.
2023-11-14 14:34:26,389 - stpipe.HighLevelPipeline.resample - INFO - ResampleStep instance created.
2023-11-14 14:34:26,389 - stpipe - INFO - Hostname: attilla.fios-router.home.stsci.edu
2023-11-14 14:34:26,389 - stpipe - INFO - OS: Darwin
2023-11-14 14:34:26,457 - stpipe.HighLevelPipeline - INFO - Step HighLevelPipeline running with args ('r00005-a3001_image_001_short_asn.json',).
2023-11-14 14:34:26,460 - stpipe.HighLevelPipeline - INFO - Step HighLevelPipeline parameters are: {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.asdf', 'output_use_model': False, 'output_use_index': True, 'save_results': True, 'skip': False, 'suffix': None, 'search_output_file': True, 'input_dir': '', 'steps': {'skymatch': {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.asdf', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': True, 'input_dir': '', 'skymethod': 'match', 'match_down': True, 'subtract': False, 'stepsize': None, 'skystat': 'mode', 'dqbits': '~DO_NOT_USE+NON_SCIENCE', 'lower': None, 'upper': None, 'nclip': 5, 'lsigma': 4.0, 'usigma': 4.0, 'binwidth': 0.1}, 'outlier_detection': {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.asdf', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': True, 'input_dir': ''}, 'resample': {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.asdf', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': True, 'input_dir': '', 'pixfrac': 1.0, 'kernel': 'square', 'fillval': 'INDEF', 'weight_type': 'ivm', 'output_shape': None, 'crpix': None, 'crval': None, 'rotation': None, 'pixel_scale_ratio': 1.0, 'pixel_scale': None, 'output_wcs': '', 'single': False, 'blendheaders': True, 'allowed_memory': None, 'in_memory': True}}}
2023-11-14 14:34:26,460 - stpipe.HighLevelPipeline - INFO - Starting Roman high level calibration pipeline ...
2023-11-14 14:34:26,530 - stpipe.HighLevelPipeline.skymatch - INFO - Step skymatch running with args ('r00005-a3001_image_001_short_asn.json',).
2023-11-14 14:34:26,531 - stpipe.HighLevelPipeline.skymatch - INFO - Step skymatch parameters are: {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.asdf', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': 'skymatch', 'search_output_file': True, 'input_dir': '', 'skymethod': 'match', 'match_down': True, 'subtract': False, 'stepsize': None, 'skystat': 'mode', 'dqbits': '~DO_NOT_USE+NON_SCIENCE', 'lower': None, 'upper': None, 'nclip': 5, 'lsigma': 4.0, 'usigma': 4.0, 'binwidth': 0.1}
2023-11-14 14:34:27,064 - stpipe.HighLevelPipeline.skymatch - WARNING - /Users/ddavis/miniconda3/envs/rcal_dev/lib/python3.11/site-packages/erfa/core.py:154: ErfaWarning: ERFA function "dtf2d" yielded 1 of "dubious year (Note 6)"
2023-11-14 14:34:27,064 - stpipe.HighLevelPipeline.skymatch - WARNING -   warnings.warn('ERFA function "{}" yielded {}'.format(func_name, wmsg),
2023-11-14 14:34:27,064 - stpipe.HighLevelPipeline.skymatch - WARNING -
2023-11-14 14:34:27,693 - stpipe.HighLevelPipeline.skymatch - INFO -
2023-11-14 14:34:27,693 - stpipe.HighLevelPipeline.skymatch - INFO - ***** romancal.skymatch.skymatch.match() started on 2023-11-14 14:34:27.692978
2023-11-14 14:34:27,693 - stpipe.HighLevelPipeline.skymatch - INFO -
2023-11-14 14:34:27,693 - stpipe.HighLevelPipeline.skymatch - INFO - Sky computation method: 'match'
2023-11-14 14:34:27,693 - stpipe.HighLevelPipeline.skymatch - INFO - Sky matching direction: DOWN
2023-11-14 14:34:27,693 - stpipe.HighLevelPipeline.skymatch - INFO - Sky subtraction from image data: OFF
2023-11-14 14:34:27,693 - stpipe.HighLevelPipeline.skymatch - INFO -
2023-11-14 14:34:27,693 - stpipe.HighLevelPipeline.skymatch - INFO - ----  Computing differences in sky values in overlapping regions.
2023-11-14 14:34:29,811 - stpipe.HighLevelPipeline.skymatch - INFO -    *  Image ID=None. Sky background: 0 electron / s
2023-11-14 14:34:29,812 - stpipe.HighLevelPipeline.skymatch - INFO -    *  Image ID=None. Sky background: 0.836698 electron / s
2023-11-14 14:34:29,812 - stpipe.HighLevelPipeline.skymatch - INFO -
2023-11-14 14:34:29,812 - stpipe.HighLevelPipeline.skymatch - INFO - ***** romancal.skymatch.skymatch.match() ended on 2023-11-14 14:34:29.812262
2023-11-14 14:34:29,812 - stpipe.HighLevelPipeline.skymatch - INFO - ***** romancal.skymatch.skymatch.match() TOTAL RUN TIME: 0:00:02.119284
2023-11-14 14:34:29,812 - stpipe.HighLevelPipeline.skymatch - INFO -
2023-11-14 14:34:29,818 - stpipe.HighLevelPipeline.skymatch - INFO - Step skymatch done
2023-11-14 14:34:29,900 - stpipe.HighLevelPipeline.resample - INFO - Step resample running with args ('r00005-a3001_image_001_short_asn.json',).
2023-11-14 14:34:29,902 - stpipe.HighLevelPipeline.resample - INFO - Step resample parameters are: {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.asdf', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': True, 'input_dir': '', 'pixfrac': 1.0, 'kernel': 'square', 'fillval': 'INDEF', 'weight_type': 'ivm', 'output_shape': None, 'crpix': None, 'crval': None, 'rotation': None, 'pixel_scale_ratio': 1.0, 'pixel_scale': None, 'output_wcs': '', 'single': False, 'blendheaders': True, 'allowed_memory': None, 'in_memory': True}
2023-11-14 14:34:30,486 - stpipe.HighLevelPipeline.resample - INFO - Setting drizzle's default parameters...
2023-11-14 14:34:30,487 - stpipe.HighLevelPipeline.resample - INFO -   using: pixfrac=1.0
2023-11-14 14:34:30,487 - stpipe.HighLevelPipeline.resample - INFO -   using: kernel='square'
2023-11-14 14:34:30,487 - stpipe.HighLevelPipeline.resample - INFO -   using: fillval='INDEF'
2023-11-14 14:34:30,488 - stpipe.HighLevelPipeline.resample - INFO -   using: wht_type='ivm'
2023-11-14 14:34:30,488 - stpipe.HighLevelPipeline.resample - INFO -   using: pscale_ratio=1.0
2023-11-14 14:34:30,488 - stpipe.HighLevelPipeline.resample - INFO - Driz parameter kernel: square
2023-11-14 14:34:30,488 - stpipe.HighLevelPipeline.resample - INFO - Driz parameter pixfrac: 1.0
2023-11-14 14:34:30,488 - stpipe.HighLevelPipeline.resample - INFO - Driz parameter fillval: INDEF
2023-11-14 14:34:30,488 - stpipe.HighLevelPipeline.resample - INFO - Driz parameter weight_type: ivm
2023-11-14 14:34:30,488 - stpipe.HighLevelPipeline.resample - INFO - Output pixel scale ratio: 1.0
2023-11-14 14:34:30,902 - stpipe.HighLevelPipeline.resample - WARNING - /Users/ddavis/miniconda3/envs/rcal_dev/lib/python3.11/site-packages/erfa/core.py:154: ErfaWarning: ERFA function "d2dtf" yielded 1 of "dubious year (Note 5)"
2023-11-14 14:34:30,902 - stpipe.HighLevelPipeline.resample - WARNING -   warnings.warn('ERFA function "{}" yielded {}'.format(func_name, wmsg),
2023-11-14 14:34:30,902 - stpipe.HighLevelPipeline.resample - WARNING -
2023-11-14 14:34:30,907 - stpipe.HighLevelPipeline.resample - WARNING - /Users/ddavis/miniconda3/envs/rcal_dev/lib/python3.11/site-packages/erfa/core.py:154: ErfaWarning: ERFA function "d2dtf" yielded 1 of "dubious year (Note 5)"
2023-11-14 14:34:30,907 - stpipe.HighLevelPipeline.resample - WARNING -   warnings.warn('ERFA function "{}" yielded {}'.format(func_name, wmsg),
2023-11-14 14:34:30,907 - stpipe.HighLevelPipeline.resample - WARNING -
2023-11-14 14:34:31,297 - stpipe.HighLevelPipeline.resample - WARNING - /Users/ddavis/miniconda3/envs/rcal_dev/lib/python3.11/site-packages/erfa/core.py:154: ErfaWarning: ERFA function "d2dtf" yielded 1 of "dubious year (Note 5)"
2023-11-14 14:34:31,297 - stpipe.HighLevelPipeline.resample - WARNING -   warnings.warn('ERFA function "{}" yielded {}'.format(func_name, wmsg),
2023-11-14 14:34:31,297 - stpipe.HighLevelPipeline.resample - WARNING -
2023-11-14 14:34:31,300 - stpipe.HighLevelPipeline.resample - WARNING - /Users/ddavis/miniconda3/envs/rcal_dev/lib/python3.11/site-packages/erfa/core.py:154: ErfaWarning: ERFA function "d2dtf" yielded 1 of "dubious year (Note 5)"
2023-11-14 14:34:31,300 - stpipe.HighLevelPipeline.resample - WARNING -   warnings.warn('ERFA function "{}" yielded {}'.format(func_name, wmsg),
2023-11-14 14:34:31,300 - stpipe.HighLevelPipeline.resample - WARNING -
2023-11-14 14:34:31,723 - stpipe.HighLevelPipeline.resample - INFO - Skipping blendheaders for now.
2023-11-14 14:34:31,735 - stpipe.HighLevelPipeline.resample - INFO - Resampling science data
2023-11-14 14:34:39,102 - stpipe.HighLevelPipeline.resample - INFO - Drizzling (4088, 4088) --> (5003, 4358)
2023-11-14 14:34:43,925 - stpipe.HighLevelPipeline.resample - INFO - Results from cdrizzle.tdriz():             '_vers'=Callable C-based DRIZZLE Version 1.12 (28th June 2018), 'nmiss'=-4088, 'nskip'=0
2023-11-14 14:34:50,513 - stpipe.HighLevelPipeline.resample - INFO - Drizzling (4088, 4088) --> (5003, 4358)
2023-11-14 14:34:55,273 - stpipe.HighLevelPipeline.resample - INFO - Results from cdrizzle.tdriz():             '_vers'=Callable C-based DRIZZLE Version 1.12 (28th June 2018), 'nmiss'=-4058, 'nskip'=0
2023-11-14 14:34:55,306 - stpipe.HighLevelPipeline.resample - INFO - Resampling var_rnoise
2023-11-14 14:35:01,954 - stpipe.HighLevelPipeline.resample - INFO - Drizzling (4088, 4088) --> (5003, 4358)
2023-11-14 14:35:13,729 - stpipe.HighLevelPipeline.resample - INFO - Drizzling (4088, 4088) --> (5003, 4358)
2023-11-14 14:35:18,807 - stpipe.HighLevelPipeline.resample - INFO - Resampling var_poisson
2023-11-14 14:35:25,536 - stpipe.HighLevelPipeline.resample - INFO - Drizzling (4088, 4088) --> (5003, 4358)
2023-11-14 14:35:37,467 - stpipe.HighLevelPipeline.resample - INFO - Drizzling (4088, 4088) --> (5003, 4358)
2023-11-14 14:35:42,687 - stpipe.HighLevelPipeline.resample - INFO - Resampling var_flat
2023-11-14 14:35:49,194 - stpipe.HighLevelPipeline.resample - INFO - Drizzling (4088, 4088) --> (5003, 4358)
2023-11-14 14:36:00,585 - stpipe.HighLevelPipeline.resample - INFO - Drizzling (4088, 4088) --> (5003, 4358)
2023-11-14 14:36:05,466 - stpipe.HighLevelPipeline.resample - INFO - Resampling exposure time
2023-11-14 14:36:12,117 - stpipe.HighLevelPipeline.resample - INFO - Drizzling (4088, 4088) --> (5003, 4358)
2023-11-14 14:36:23,781 - stpipe.HighLevelPipeline.resample - INFO - Drizzling (4088, 4088) --> (5003, 4358)
2023-11-14 14:36:29,153 - stpipe.HighLevelPipeline.resample - INFO - Mean, max exposure times: 285.8, 285.8
2023-11-14 14:36:29,160 - stpipe.HighLevelPipeline.resample - WARNING - /Users/ddavis/miniconda3/envs/rcal_dev/lib/python3.11/site-packages/erfa/core.py:154: ErfaWarning: ERFA function "d2dtf" yielded 1 of "dubious year (Note 5)"
2023-11-14 14:36:29,160 - stpipe.HighLevelPipeline.resample - WARNING -   warnings.warn('ERFA function "{}" yielded {}'.format(func_name, wmsg),
2023-11-14 14:36:29,160 - stpipe.HighLevelPipeline.resample - WARNING -
2023-11-14 14:36:29,248 - stpipe.HighLevelPipeline.resample - INFO - Update S_REGION to POLYGON ICRS  80.021216639 29.980879920 79.869310622 30.056876908 79.793002337 29.942238797 79.944791638 29.866329380
2023-11-14 14:36:29,304 - stpipe.HighLevelPipeline.resample - INFO - Step resample done
2023-11-14 14:36:29,529 - stpipe.HighLevelPipeline - WARNING - /Users/ddavis/miniconda3/envs/rcal_dev/lib/python3.11/site-packages/erfa/core.py:154: ErfaWarning: ERFA function "d2dtf" yielded 1 of "dubious year (Note 5)"
2023-11-14 14:36:29,530 - stpipe.HighLevelPipeline - WARNING -   warnings.warn('ERFA function "{}" yielded {}'.format(func_name, wmsg),
2023-11-14 14:36:29,530 - stpipe.HighLevelPipeline - WARNING -
2023-11-14 14:36:29,534 - stpipe.HighLevelPipeline - WARNING - /Users/ddavis/miniconda3/envs/rcal_dev/lib/python3.11/site-packages/erfa/core.py:154: ErfaWarning: ERFA function "d2dtf" yielded 1 of "dubious year (Note 5)"
2023-11-14 14:36:29,534 - stpipe.HighLevelPipeline - WARNING -   warnings.warn('ERFA function "{}" yielded {}'.format(func_name, wmsg),
2023-11-14 14:36:29,534 - stpipe.HighLevelPipeline - WARNING -
2023-11-14 14:36:29,601 - stpipe.HighLevelPipeline - WARNING - /Users/ddavis/miniconda3/envs/rcal_dev/lib/python3.11/site-packages/erfa/core.py:154: ErfaWarning: ERFA function "d2dtf" yielded 1 of "dubious year (Note 5)"
2023-11-14 14:36:29,601 - stpipe.HighLevelPipeline - WARNING -   warnings.warn('ERFA function "{}" yielded {}'.format(func_name, wmsg),
2023-11-14 14:36:29,601 - stpipe.HighLevelPipeline - WARNING -
2023-11-14 14:36:29,604 - stpipe.HighLevelPipeline - WARNING - /Users/ddavis/miniconda3/envs/rcal_dev/lib/python3.11/site-packages/erfa/core.py:154: ErfaWarning: ERFA function "d2dtf" yielded 1 of "dubious year (Note 5)"
2023-11-14 14:36:29,604 - stpipe.HighLevelPipeline - WARNING -   warnings.warn('ERFA function "{}" yielded {}'.format(func_name, wmsg),
2023-11-14 14:36:29,604 - stpipe.HighLevelPipeline - WARNING -
2023-11-14 14:36:30,792 - stpipe.HighLevelPipeline - INFO - Saved model in r00005-a3001_image_001_short_asn_highlevelpipeline.asdf
2023-11-14 14:36:30,792 - stpipe.HighLevelPipeline - INFO - Step HighLevelPipeline done
...
Note: outlier detection is still being skipped and the output filename is a bit crazy. 
We have: r00005-a3001_image_001_short_asn_highlevelpipeline.asdf
I'd prefer something like,
r0000501001001001001_WFI01_<filter>_i2d.asdf

**Checklist**
- [x] added entry in `CHANGES.rst` under the corresponding subsection
- [N/A] updated relevant tests
- [N/A] updated relevant documentation
- [x] updated relevant milestone(s)
- [x] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below. [How to run regression tests on a PR](https://github.com/spacetelescope/romancal/wiki/Running-Regression-Tests-Against-PR-Branches)
 Regression tests will fail until we have new files from INS

### CODE REVIEW

File: romancal/pipeline/highlevel_pipeline.py
Comment (schlafly) [inline romancal/pipeline/highlevel_pipeline.py:101]: I'm confused here. I think skymatching & outlier detection need to get the full list of exposures being combined, and this looks like instead they get them one by one. I don't think skymatch & outlier detection can do anything with a single exposure, and presumably just return it back ~unchanged. I think I would have guessed that these would get sent the full asn / ModelContainer, and there wouldn't be a loop in the HLP?

Comment (ddavis-stsci) [reply romancal/pipeline/highlevel_pipeline.py:101]: I think you are correct but at the moment I have no data for testing. So for the interim I've keep the elp behavior . My impression was that Nadia wanted this in so I went ahead and plan on fixing any issues once we have working examples. I have no problem waiting until we have input data to merge this but we'll be quite close to the release deadline.

File: romancal/pipeline/highlevel_pipeline.py
Comment (nden) [inline romancal/pipeline/highlevel_pipeline.py:26]:
```suggestion
    ``skymatch``, ``outlier_detection`` and ``resample``.
```

**Comment by nden** [comment]
In addition to the inline comments, I looked at the module `highlevel_pipeline.py` (sorry for the late review).
I don't understand the setting of suffix if `filetype` is `asn`. Different suffixes are set always to the skymatch step. Is this an oversight?

In addition the `setup_output` function is not used anywhere? Is it intended to be used?

File: romancal/pipeline/highlevel_pipeline.py
Comment (nden) [reply romancal/pipeline/highlevel_pipeline.py]: We have input data, the data used for testing resample. It's a FOV dithered simulation and Mairan knows the location.

**Comment by nden** [comment]
In the case of an association file, the code reads it into a `ModelContainer` but then uses the original `input` as input to the following steps. Was this intended to be modified later?

File: romancal/pipeline/__init__.py
Comment (schlafly) [reply romancal/pipeline/__init__.py:8]: You may have a better picture than I do, but L4 seems to be to be a catchall for "everything after L3."  It includes single file catalog products, multi-band catalog products, difference images, difference image catalogs, forced photometry, ....  IMO we'll end up with several different things that construct the different kinds of L4 products.

File: romancal/pipeline/highlevel_pipeline.py
Comment (ddavis-stsci) [reply romancal/pipeline/highlevel_pipeline.py:38]: Updated

File: romancal/pipeline/__init__.py
Comment (ddavis-stsci) [reply romancal/pipeline/__init__.py:8]: For now I was just matching what I saw in the design document (https://innerspace.stsci.edu/display/RDMSDOC/DMS+Operational+Design) which has l3 and l4 products described. If we need to break this into roman_l3 and roman_l4 that should not be an issue from the coding point of view.

**Review by schlafly — APPROVED** [review]
Let's defer further improvements to B13; approved.

**Pull request merged**
